### ISSUE TITLE
Ordnance Survey Coordinate System (OSGB)

### ISSUE THREAD

**Issue by Stephen-A-J** [osmandapp/OsmAnd#11299]
Request suppot of the OSGB coordinate system as an alternative to Lat/Long. This is the standard system used in the UK.

This feature is supported in most standalone GPS units and Android mapping apps that I have seen.

In the UK most (paper) walking books and magazines list the waypoints on a route using the OS (Ordnance Survey) grid ref. This is quite short (eg "ST 123 678") and so is very easy to type in. This ref is accurate to 100m which is good enough for most purposes, but a longer form "ST 12345 67890" gives an accuracy of 1m.

Workaround: At the moment I have to type these waypoints into another mapping program, save the generated route as a GPX file and then load that into OSMAND. It would obviously be great if I could enter it into OSMAND direct.

**Comment by stevenculshaw** [comment]
+1 for this, if you're packing printed maps for planning or backup it would be nice if the two matched so you don't need to convert all your waypoints which is tedious.

**Comment by nickleverton** [comment]
Yes another +1 please. I've been looking for a plugin to do this without success.to support OSGB NGR natively would improve its useability massively since all UK location references use it.

**Comment by wookey** [comment]
Another request. OSGB is still heavily used in the UK. I am on a group walking holiday right now and the walk starts are given as OSGB and What3words (boo!). Not Lat/long or OLC.
I realise that there are a lot of national co-ordinate systems, with varying degrees of popularity and putting them all in the list would make a very long list, but a plugin for this functionality, or some other way to select _which_ national co-ordinate system you wanted shown in the list would be marvellous.
As I just failed to find anything that does OSGB->something Osmand understands conversion, I don't mind coding up something (in principle - I know nothing about the osmand plugin architecture or what language these things are written in, but I'm going to presume it's not _that_ hard).
We should perhaps discuss here what the right interface for choosing national co-ordinate systems is. Do we just keep making the co-ordinate entry list longer and hope that there aren't requests for too many, or arrange to include one (or two?) national systems on the list. If the latter where does that configuration happen?

**Comment by mattgreenuk** [comment]
Another +1 from me :-)
The OS reference system is used almost exclusively across the UK - we rarely see lat/long or other coordinate systems here. There is the broader question of should the list of supported coordinate systems just grow and grow? Some are country-specific whereas some are global - should they appear the same or as a hierarchy in the selection menu?

**Comment by Scibearspace** [comment]
Hello All,

Just found out I am at an event that uses the OS grid reference for the start point / event parking. So I needed to add this into OSMand so as I don't get lost on the way there !

So I found a site that does this conversion :https://gridreferencefinder.com/

When you then add in the OS grid refe the 'search term' comes up as follows in the search bar : 
https://gridreferencefinder.com/#gr=TQ6577245588|TQ_s_65772_s_45588|1

The OS grid reference I entered was TQ6577245588.

It then displays a pop up of all the other various coordinates that are avalible, obviously for OSM the most interesting is the Lat/lon variants.

They also enable batch conversion (see page : https://gridreferencefinder.com/batchConvert/batchConvert.php) so they are clearly using php in their code (not a big surprise).

I'm going to keep investigating to see if I can find the 'actual' script for the converstion, and I'll post it here. As one of the other posters has already said, I don't know what script language you use in the app / on the web page.

edit 1 :
Also possibly usefull is the the FAQ 7 : https://gridreferencefinder.com/faq.php#How_Do_I_Construct_My_Own_Multi_Point_URL_Link_to_a_Map

Which shows the code to create a point, which will then return the same info. The question is how to best capture / parse that returned information.

Edit 2 : 

found 2 files that are usefull on the www.ordnancesurvey.co.uk site, these are a pair of XL files (with macros, so VB) that convert OS coordinates into regular GPS coordinates (and a whole bunch of other stuff). I won't paste the file here directly but rather the links to them : 
https://www.ordnancesurvey.co.uk/documents/resources/projection-and-transformation-calculations.xls
https://www.ordnancesurvey.co.uk/documents/resources/maptile-gridref-conversion.xlsm

I'm gradually working my way through them, it seems like a more complicated process than I had originally expected. Although the macro code is quite understandable / readable.

**Comment by wookey** [comment]
I was at another event where missing OSGB co-ordinates in OSMand was a problem, and I see this hasn't progressed in the last 4 years. 
I guess someone needs to actually provide a patch?
Proj can do these conversions, and I presume OSMand already uses proj for conversions. So adding more input and output modes should be straighforward. 

With regard to the list growing forever, we can just provide configuration to let people select which ones they want. The default would ideally show 'standard global formats' (lat/long and UTM) plus co-ordinate systems used in the current location and configured country of origin. Not sure how easy that is to do. UK users definitely want OSGB. There are local grid for France (Lambert), the US (NAD) and so on, but we'd need a list of which are still used/useful where. That can be generated by requests like this issue, once there is a way to fulfill those requests.

**Comment by rKsanu2MMYvypWePtQWM** [comment]
I'd also would like to chime in, having a menu with custom coordinate formats as well as ability to define custom formats (like in Garmin GPS receivers) would be great. Honesty was surprised when I looked thru the options and didn't find this feature already...
